File: backend/app/models/workout.py
from datetime import datetime, UTC

from sqlalchemy import JSON, Column, Integer, DateTime, ForeignKey
from sqlalchemy.orm import relationship

from app.database import Base


class WorkoutSession(Base):
    __tablename__ = "workout_sessions"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    exercise_type_id = Column(Integer, ForeignKey("exercise_types.id"), nullable=False)
    rep_count = Column(Integer, nullable=False)
    date = Column(DateTime, default=lambda: datetime.now(UTC))
    average_quality_score = Column(Integer, nullable=True)
    analysis = Column(JSON, nullable=True)

    user = relationship("User", back_populates="workouts")
    exercise_type = relationship("ExerciseType", back_populates="workouts")

    @property
    def exercise_type_code(self) -> str:
        return self.exercise_type.code

File: flutter_app/lib/models/workout.dart
import 'workout_analysis.dart';

class Workout {
  final int id;
  final int userId;
  final String exerciseType;
  final int repCount;
  final DateTime date;
  final int? averageQualityScore;
  final WorkoutAnalysis? analysis;

  Workout({
    required this.id,
    required this.userId,
    required this.exerciseType,
    required this.repCount,
    required this.date,
    this.averageQualityScore,
    this.analysis,
  });

  factory Workout.fromJson(Map<String, dynamic> json) {
    return Workout(
      id: json['id'],
      userId: json['user_id'],
      exerciseType: json['exercise_type'],
      repCount: json['rep_count'],
      date: DateTime.parse(json['date']),
      averageQualityScore: json['average_quality_score'] as int?,
      analysis: json['analysis'] == null
          ? null
          : WorkoutAnalysis.fromJson(
              Map<String, dynamic>.from(json['analysis'] as Map),
            ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'exercise_type': exerciseType,
      'rep_count': repCount,
      if (averageQualityScore != null) 'average_quality_score': averageQualityScore,
      if (analysis != null) 'analysis': analysis!.toJson(),
    };
  }
}
